src/sarif-converter-21.ts

```diff
@@ -0,0 +1,338 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+import * as Axe from 'axe-core';
+import { getAxeToolProperties21 } from './axe-tool-property-provider-21';
+import { ConverterOptions } from './converter-options';
+import { getConverterProperties } from './converter-property-provider';
+import {
+    DecoratedAxeResult,
+    DecoratedAxeResults,
+} from './decorated-axe-results';
+import { DictionaryStringTo } from './dictionary-types';
+import { EnvironmentData } from './environment-data';
+import { getEnvironmentDataFromResults } from './environment-data-provider';
+import { getInvocations21 } from './invocation-provider-21';
+import * as CustomSarif from './sarif/custom-sarif-types-21';
+import * as Sarif from './sarif/sarif-2.1.2';
+import { SarifLog21 } from './sarif/sarif-log-21';
+import { isNotEmpty } from './string-utils';
+
+export function defaultSarifConverter21(): SarifConverter21 {
+    return new SarifConverter21(
+        getConverterProperties,
+        getAxeToolProperties21,
+        getInvocations21,
+    );
+}
+export class SarifConverter21 {
+    public constructor(
+        private getConverterToolProperties: () => Sarif.Run['conversion'],
+        private getAxeProperties: () => Sarif.Run['tool'],
+        private invocationConverter: (
+            environmentData: EnvironmentData,
+        ) => Sarif.Invocation[],
+    ) {}
+
+    public convert(
+        results: DecoratedAxeResults,
+        options: ConverterOptions,
+    ): SarifLog21 {
+        return {
+            version: CustomSarif.SarifLogVersion21.v21,
+            runs: [this.convertRun(results, options)],
+        };
+    }
+
+    private convertRun(
+        results: DecoratedAxeResults,
+        options: ConverterOptions,
+    ): Sarif.Run {
+        const files: DictionaryStringTo<Sarif.File> = {};
+        files[results.targetPageUrl] = {
+            mimeType: 'text/html',
+            properties: {
+                tags: ['target'],
+                title: results.targetPageTitle,
+            },
+        };
+
+        let properties: DictionaryStringTo<string> = {};
+
+        if (options && options.scanName !== undefined) {
+            properties = {
+                scanName: options.scanName,
+            };
+        }
+
+        const run: Sarif.Run = {
+            conversion: this.getConverterToolProperties(),
+            tool: this.getAxeProperties(),
+            invocations: this.invocationConverter(
+                getEnvironmentDataFromResults(results),
+            ),
+            files: files,
+            results: this.convertResults(results, properties),
+            resources: {
+                rules: this.convertResultsToRules(results),
+            },
+            properties: {},
```

Comment:
As someone new to the code it seems odd that the value of `properties` is an empty dict when we define `properties` above. If it is intended, can we consider renaming the variable to disambiguate?

Reply:
I agree that it is odd for the `properties` property of the Sarif.Run object to be an empty object (it was present in the v2.0.0 converter as well). Since it isn't being populated with any useful information and is not present in the sample output for v2.1.2, I will remove it in a future PR.